### PR TITLE
Mail Notification Settings UI & Mail Sender Library Upgrade

### LINKED ISSUE #395: SMTP sending issues
https://forums.bluecherrydvr.com/t/no-mails-can-be-sent-using-authentication/299

### PR DESCRIPTION
**Pull request by curtishall**
- Enhanced mail notification settings UI
- Obsolete pear mail library dependency replaced to swiftmailer library

Resolves #395

### CODE REVIEW

**Comment by antonsviridenko** [comment]
@tugrul 

what is 
`require_once 'vendor/autoload.php';`

where this file is expected to be sourced from?
Bluecherry web UI is packaged into the .deb package and therefore all dependencies should be bundled into the package or be present in Ubuntu LTS repository

**Comment by tugrul** [comment]
@antonsviridenko It is composer's autoloader script and It is going to present when you run `composer install` in www directory.

This is decent way to manage dependencies of php projects.

You can put install hooks to deb package or you can install dependency before packaging.

**Comment by curtishall** [comment]
@antonsviridenko php-swiftmailer exists as a deb package if that is easier

**Comment by tugrul** [comment]
@curtishall `php-swiftmailer` package is too old.  https://github.com/jbouse-debian/php-swiftmailer

composer is best way to use fresh version of php dependencies.

**Comment by antonsviridenko** [comment]
@tugrul actually I just checked and in all distros that we support there is a latest version 5.4.2 of swiftmailer packaged, and it was released 5 years ago

https://packages.ubuntu.com/search?keywords=php-swiftmailer&searchon=names&suite=bionic&section=all
https://packages.ubuntu.com/search?keywords=php-swiftmailer&searchon=names&suite=focal&section=all
https://packages.debian.org/stable/php/php-swiftmailer

so there is no reason to use compose here when distro package exists in repository

**Comment by antonsviridenko** [comment]
Or it's not the latest one?

**Comment by antonsviridenko** [comment]
I guess this one is the new upstream now, right?
https://github.com/swiftmailer/swiftmailer

**Comment by antonsviridenko** [comment]
Can you provide instructions how to install swiftmailer using composer then? If we decide to bundle dependencies into the .deb file

**Comment by tugrul** [comment]
@antonsviridenko yes it is official repository and fresh one. Using fresh version is good idea to avoid possible bugs and security holes.

Using composer is very easy I'm going to write common steps below.

You can follow instructions of [this page](https://getcomposer.org/download/) to install composer cli tool if you didn't install before.

You are going to obtain `composer.phar` file in current working directory. It is a runnable php archive file and you can execute it like `./composer.phar`. You can run it in place or you can move it (`mv composer.phar /usr/local/bin/composer`) to bin directory to use globally.

Second step is switching to bluecherry's www directory `cd /path/to/bluecherry/www/` then executing `composer install` cli command. It'is going to read required dependencies `composer.json` and download dependencies inside  `/path/to/bluecherry/www/vendor/` folder.

It is ready to use 👍

**Comment by antonsviridenko** [comment]
I tried to send test email on a non-configured system (no email settings, no sendmail installed in the system), gives these errors in Apache logs:
```
[Fri Jan 08  05:25:58.659820 2021] [php7:error] [pid 13173] [client 10.0.2.2:54414]  PHP Fatal error:  Uncaught Swift_RfcComplianceException: Address in  mailbox given [] does not comply with RFC 2822, 3.6.2. in  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/Headers/MailboxHeader.php:355
Stack  trace:
#0  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/Headers/MailboxHeader.php(272):  Swift_Mime_Headers_MailboxHeader->assertValidAddress('')
#1  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/Headers/MailboxHeader.php(117):  Swift_Mime_Headers_MailboxHeader->normalizeMailboxes(Array)
#2  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/Headers/MailboxHeader.php(74):  Swift_Mime_Headers_MailboxHeader->setNameAddresses(Array)
#3  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/SimpleHeaderFactory.php(61):  Swift_Mime_Headers_MailboxHeader->setFieldBodyModel(Array)
#4  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/SimpleHeaderSet.php(  in  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Mime/Headers/MailboxHeader.php  on line 355, referer: https://127.0.0.1:7001/general
--
[Fri  Jan 08 05:26:07.234047 2021] [php7:error] [pid 12702] [client  10.0.2.2:54432] PHP Fatal error:  Uncaught Swift_TransportException:  Expected response code 220 but got an empty response\
og data:
++  Starting Swift_SendmailTransport
!! Expected response code 220 but got  an empty response (code: 0) in  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Plugins/LoggerPlugin.php:124
Stack  trace:
#0  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Events/SimpleEventDispatcher.php(138):   Swift_Plugins_LoggerPlugin->exceptionThrown(Object(Swift_Events_TransportExceptionEvent))
#1   /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Events/SimpleEventDispatcher.php(116):  Swift_Events_SimpleEventDispatcher->bubble(Array,  Object(Swift_Events_TransportExceptionEvent), 'exceptionThrown')
#2  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Transport/AbstractSmtpTransport.php(435):   Swift_Events_SimpleEventDispatcher->dispatchEvent(Object(Swift_Events_TransportExceptionEvent),  'exceptionThrown')
#3 /usr/share/bluecherry/www/v in  /usr/share/bluecherry/www/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Plugins/LoggerPlugin.php  on line 124, referer: https://127.0.0.1:7001/general

```
and displays 'err ajax' message box in web UI. I think it would be nice to handle these exceptions and display more meaningful error messages to user.

**Comment by tugrul** [comment]
Because sender email address is empty. It is going to send if you set from address to mail sender configuration on dashboard. I made small change for this. It will not set from address if there is no configuration for this.

**Comment by antonsviridenko** [comment]
I think we need some more improvements,
I propose to split "SSL: None, TLS/SSL" dropdown list to 3 possible options:
**Security**:

- TLS
- SSL
- None

And make "TLS" default option selected.
Also it is better to have dropdowns for SMTP port number with values 25, 465, 587.

I've just tried to send test email through GMail and it does not accept less secure authentication methods.

File: www/ajax/general.php [hunk 1]
```diff
@@ -16,42 +17,53 @@ public function getData()
...
+                $transport = (new \Swift_SmtpTransport($global_settings->data['G_SMTP_HOST']))
+                    ->setUsername($global_settings->data['G_SMTP_USERNAME'])
+                    ->setPassword($global_settings->data['G_SMTP_PASSWORD']);
+
+                if (isset($global_settings->data['G_SMTP_PORT'])) {
+                    $transport->setPort($global_settings->data['G_SMTP_PORT']);
+                }
+
+                if (isset($global_settings->data['G_SMTP_SSL']) && $global_settings->data['G_SMTP_SSL'] === 'tls') {
+                    $transport->setEncryption('ssl');
```
Comment (antonsviridenko) [inline www/ajax/general.php:51]: Let's have 'tls' here by default. Swift supports  both 'ssl' and 'tls' values here.

File: www/template/ajax/general.php [hunk 2]
```diff
@@ -152,84 +152,95 @@
...
-                </select>
+                <div class="row">
+                    <div class="col-lg-5">
+                        <select class="form-control" name='G_SMTP_SERVICE'>
+                            <option value='default' <?php echo ($global_settings->data['G_SMTP_SERVICE']=='default') ? 'selected': '';?>><?php echo G_SMTP_DEF_MTA; ?></option>
+                            <option value='smtp' <?php echo ($global_settings->data['G_SMTP_SERVICE']=='smtp') ? 'selected': '';?>><?php echo G_SMTP_SMTP; ?></option>
+                        </select>
+                        <p class='help-block'><small><?php echo G_SMTP_SERVICE_EX; ?></small></p>
+                    </div>
+                    <label class="control-label col-lg-2"><?php echo G_SMTP_SSL; ?></label>
```
Comment (antonsviridenko) [inline www/template/ajax/general.php:166]: Let's change label here to something like "Security" or "Transport Security" or "Security Options"

File: www/template/ajax/general.php [hunk 3]
```diff
@@ -152,84 +152,95 @@
...
+                    <div class="col-lg-5">
+                        <select class="form-control" name='G_SMTP_SSL'>
+                            <option value='none' <?php echo ($global_settings->data['G_SMTP_SSL']=='none') ? 'selected': '';?>><?php echo L_NONE; ?></option>
+                            <option value='tls' <?php echo ($global_settings->data['G_SMTP_SSL']=='tls') ? 'selected': '';?>>TLS/SSL</option>
```
Comment (antonsviridenko) [inline www/template/ajax/general.php:170]: Let's split this and make two separate options "TLS" and "SSL"

**Review by antonsviridenko — CHANGES_REQUESTED** [review]

**Comment by tugrul** [comment]
I pushed changes. Dropdown for port numbers is not good idea because standard ports not available some cases. Maybe user should use 2525 instead of 25 or something else. Some IT infrastructures not allows to use standard ports and they have to use some workaround like port forwarding or virtual port numbers.

**Review by antonsviridenko — APPROVED** [review]
ok

**Pull request merged**